instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/RedisConnectionProviderInstrumentation.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.vertx.v4_0.redis;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import io.vertx.redis.client.RedisConnection;
+import io.vertx.redis.client.impl.RedisStandaloneConnection;
+import io.vertx.redis.client.impl.RedisURI;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+public class RedisConnectionProviderInstrumentation implements TypeInstrumentation {
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return named("io.vertx.redis.client.impl.RedisConnectionManager$RedisConnectionProvider");
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    // 4.1.0
+    transformer.applyAdviceToMethod(
+        named("init").and(not(takesArgument(0, named("io.vertx.redis.client.RedisConnection")))),
+        this.getClass().getName() + "$InitAdvice");
+    // 4.0.0
+    transformer.applyAdviceToMethod(
+        named("init").and(takesArgument(0, named("io.vertx.redis.client.RedisConnection"))),
+        this.getClass().getName() + "$InitWithConnectionAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class InitAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void onEnter(@Advice.FieldValue("redisURI") RedisURI redisUri) {
+      VertxRedisClientSingletons.setRedisUriThreadLocal(redisUri);
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void onExit() {
+      VertxRedisClientSingletons.setRedisUriThreadLocal(null);
```

Comment:
It's not immediately clear to me why 4.1.0 can use the ThreadLocal but the 4.0.0 init needs the VirtualField. Can we put a small comment about _why_?

(edit) I guess the comment in the StandAloneProviderAdvice kinda hints at this, but I was thinking about it up here.

Reply:
Can we add a comment here as well tho?